### PR TITLE
8333964: RISC-V: C2: Check "requires_strict_order" flag for floating-point add reduction

### PR DESCRIPTION
**Pull request by zifeihan**
Hi, We want to support non strictly-ordered floating-point add reduction, It was implemented by referring to RVV v1.0 [1]. please take a look and have some reviews. Thanks a lot.

We can use the Float256VectorTests.java[2] to print the Opto JIT Code, verify and observe the generation of nodes.

For example, we can use the following command to print the Opto JIT Code of a jtreg test case:
```
/home/zifeihan/jtreg/bin/jtreg \
-v:default \
-concurrency:16 -timeout:50 \
-javaoption:-XX:+UnlockExperimentalVMOptions \
-javaoption:-XX:+UseRVV \
-javaoption:-XX:+PrintOptoAssembly \
-javaoption:-XX:LogFile=/home/zifeihan/jdk/Float256VectorTests_PrintOptoAssembly.log \
-jdk:/home/zifeihan/jdk/build/linux-riscv64-server-fastdebug/jdk \
/home/zifeihan/jdk/test/jdk/jdk/incubator/vector/Float256VectorTests.java
```
We can observe the specified JIT Code log Float256VectorTests_PrintOptoAssembly.log, which contains the reduce_addF_ordered instruction for the PR implementation.
```
1e4     B28: #	out( B28 B29 ) &lt;- in( B41 B28 ) Loop( B28-B28 inner post of N2310) Freq: 98.8164
1e4     shadd  R17, R15, R10, #2	# ptr, #@shaddP_reg_reg_ext_b
1e8     addi  R17, R17, #16	# ptr, #@addP_reg_imm
1ea     loadV V1, [R17]	# vector (rvv)
1f2     reduce_addF_unordered F2, F0, V1	# KILL V2
202     fadd.s  F1, F1, F2	#@addF_reg_reg
206     addiw  R15, R15, #8	#@addI_reg_imm
208     blt  R15, R31, B28	#@cmpI_loop  P=0.500000 C=19400.000000
```
Similarly, for `reduce_addD_unordered` instruction, we can use the `test/jdk/jdk/incubator/vector/Double256VectorTests.java` test case.

### Performance testing:
FloatMaxVector.ADDLanes [2] measures the performance of add reduction for floating-point type.
Without Patch:
```
Benchmark                (size)   Mode  Cnt    Score   Error   Units
FloatMaxVector.ADDLanes    1024  thrpt    5  394.558 ± 0.044  ops/ms
```

With Patch:
```
Benchmark                (size)   Mode  Cnt    Score   Error   Units
FloatMaxVector.ADDLanes    1024  thrpt    5  627.510 ± 1.095  ops/ms
```

### Correctness testing:
 - [x] test/jdk/jdk/incubator/vector (fastdebug) qemu 8.1.50 with UseRVV
 - [x] Run tier1-3 tests on SOPHON SG2042 (release)
 - [x]  Run tier1-3 tests (release) on qemu 8.1.50 with UseRVV

[1] https://github.com/riscv/riscv-v-spec/blob/v1.0/v-spec.adoc
[2] https://github.com/openjdk/panama-vector/blob/vectorIntrinsics/test/micro/org/openjdk/bench/jdk/incubator/vector/operation/FloatMaxVector.java#L316

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8333964](https://bugs.openjdk.org/browse/JDK-8333964): RISC-V: C2: Check "requires_strict_order" flag for floating-point add reduction (**Enhancement** - P4)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/19649/head:pull/19649` \
`$ git checkout pull/19649`

Update a local copy of the PR: \
`$ git checkout pull/19649` \
`$ git pull https://git.openjdk.org/jdk.git pull/19649/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 19649`

View PR using the GUI difftool: \
`$ git pr show -t 19649`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/19649.diff">https://git.openjdk.org/jdk/pull/19649.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/19649#issuecomment-2162105208)

### CODE REVIEW

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 1]
```diff
@@ -2007,11 +2007,12 @@ instruct reduce_addL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
   ins_pipe(pipe_slow);
 %}
 
-instruct reduce_addF(fRegF dst, fRegF src1, vReg src2, vReg tmp) %{
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv_v.ad:2010]: Can you add a comment here? Like:
```
// Distinguish two cases based on requires_strict_order
// 1. Non strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
//    generated by Vector API. It is more beneficial performance-wise to do
//    an unordered FP reduction sum (vfredusum.vs).
// 2. Strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
//    generated by auto-vectorization. Must do an ordered FP reduction sum
//    (vfredosum.vs).
```

Comment (zifeihan) [reply src/hotspot/cpu/riscv/riscv_v.ad:2010]: > Can you add a comment here? Like:
> 
> ```
> // Distinguish two cases based on requires_strict_order
> // 1. Non strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
> //    generated by Vector API. It is more beneficial performance-wise to do
> //    an unordered FP reduction sum (vfredusum.vs).
> // 2. Strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
> //    generated by auto-vectorization. Must do an ordered FP reduction sum
> //    (vfredosum.vs).
> ```

Thanks for the review. Fixed.

File: src/hotspot/cpu/riscv/riscv_v.ad [hunk 2]
```diff
@@ -2007,11 +2007,19 @@ instruct reduce_addL(iRegLNoSp dst, iRegL src1, vReg src2, vReg tmp) %{
...
+// Distinguish two cases based on requires_strict_order
+// 1. Non strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
+//    generated by Vector API. It is more beneficial performance-wise to do
+//    an unordered FP reduction sum (vfredusum.vs).
+// 2. Strictly-ordered AddReductionVF/D. For example, AddReductionVF/D
+//    generated by auto-vectorization. Must do an ordered FP reduction sum
+//    (vfredosum.vs).
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv_v.ad:2016]: Nit: Can you leave an empty line after the code comment?

**Review by RealFYang — APPROVED** [review]
Looks reasonable to me.

File: src/hotspot/cpu/riscv/riscv_v.ad (hunk 2 above)
Comment (zifeihan) [reply src/hotspot/cpu/riscv/riscv_v.ad:2016]: > Nit: Can you leave an empty line after the code comment?

Fixed.

**Comment by zifeihan** [comment]
@RealFYang : Hi, Based on https://github.com/openjdk/jdk/pull/19686, I've updated some of the tests, and test case passed normally. Can you take a look? Thanks

**Review by RealFYang — APPROVED** [review]

**Comment by zifeihan** [comment]
@RealFYang : Thanks for the review.
/integrate

**Pull request closed without merging**